### PR TITLE
[IMP] account: easier payment registration for bills

### PR DESCRIPTION
**Pull request by Megaaaaaa**
Description of the issue/feature this commit addresses:

Currently, the payment registration for vendor bills is kinda hard to figure. Noticing which bills are overdue, which require a payment is not made so easy by the interface and could be improved. Also there is no way to register a payment for multiple lines at a time which could improve the UX.

---

Desired behavior after the commit is merged :

On the dashboard, all the buttons in the "Customer Invoices" and "Vendor Bills" have been renamed to remove the "Invoices" and the "Bills" inside them.

The buttons "Bills to Pay" and "Late Bills" which now are "To Pay" and "Late" (see above) redirect to the Journal Items with specific filters to show only the necessary lines.

The order of the columns in the Journal Items has been changed and their visibility has been changed too.

The default filters of the Journal Items when coming from the dashboard buttons has been changed. Payable has been replaced by Purchases.

Filters, Group by and Search by options have been added for the Due Date and the Discount Date fields in the Journal Items.

---

task-3516496

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/account/models/account_journal_dashboard.py
Comment (Megaaaaaa) [reply addons/account/models/account_journal_dashboard.py:728]: The point of putting those in the domain instead of using the already existing filters was for the search bar not to be filled with filters the user will never want to remove and therefore don't need to be displayed.
I couldn't figure a way to hide filters so I put those in the domain.

File: addons/account/views/account_journal_dashboard_view.xml
Comment (Megaaaaaa) [reply addons/account/views/account_journal_dashboard_view.xml:338]: The "Invoices to Validate", "Bills to Validate", "Unpaid Invoice" and "Late Invoice" texts are clickable and trigger the open_action method. That method assigns the right filters and domain according to the text that was clicked. However, all those redirect to an account.move view while "Bills to Pay" and "Late Bills" redirect to an account.move.line view. Using the same method would probably involve a lot of changes while using a different method is way clearer. This new method is action_open_payment_items and return an easy action which needs a simple domain change depending on whether you click on one or the other.

In short : Sales is used when clicking on an invoice related text but it is done inside the action instead of passed via the context.

File: addons/web/static/src/search/search_bar/search_bar.js
Comment (william-andre) [inline addons/web/static/src/search/search_bar/search_bar.js:1]: @odoo/rd-framework-js Are you ok with this?

The purpose is to be able to have a quick search that searches before a certain date by default (see `_search_payment_date`)

File: addons/account/views/account_move_views.xml
Comment (william-andre) [reply addons/account/views/account_move_views.xml:1668]: You resolved but it is still here?

File: addons/account/models/account_journal_dashboard.py
Comment (william-andre) [inline addons/account/models/account_journal_dashboard.py:754]: Same, should be in `account_reports`

File: addons/account/views/account_journal_dashboard_view.xml
Comment (william-andre) [inline addons/account/views/account_journal_dashboard_view.xml:324]: Is there a reason to remove the `title`?
I think it can be useful on smaller screens, for instance:
image

Could be even worse with some languages

File: addons/account/views/account_journal_dashboard_view.xml
Comment (william-andre) [inline addons/account/views/account_journal_dashboard_view.xml:335]: Any reason to add a `span`?

**Review by william-andre — COMMENTED** [review]
It's getting there! :smile:

File: addons/account/views/account_move_views.xml
Comment (Megaaaaaa) [reply addons/account/views/account_move_views.xml:1668]: I thougth your first comment was about the action_open_payment_item() method which was present in both the odoo/addons/account/models/account_journal_dashboard.py and enterprise/account_reports/models/account_aged_partner_balance.py and was indeed a duplicate. Wasn't it ?

This action_open_payment_item is the menu item in the drop-down when clicking on "Vendors" in the bar at the top of the UI. Am i not forced to create a specific action for such a button ?

File: addons/account/models/account_journal_dashboard.py
Comment (Megaaaaaa) [reply addons/account/models/account_journal_dashboard.py:742]: You want me to duplicate _get_model_info_from_id() in this model ?

File: addons/account/views/account_journal_dashboard_view.xml
Comment (Megaaaaaa) [reply addons/account/views/account_journal_dashboard_view.xml:324]: I removed them after this this comment in your first review: https://github.com/odoo/odoo/pull/138188#discussion_r1360590483 . 

I'm putting them back waiting for your confirmation but I think it is indeed better to keep them.

Comment (william-andre) [reply addons/account/views/account_journal_dashboard_view.xml:324]: Yes, I probably got confused in the diff before, which was more complicated

File: addons/account/models/account_journal_dashboard.py
Comment (william-andre) [reply addons/account/models/account_journal_dashboard.py:742]: No, I want you to extend this method in `account_reports` to use `_get_model_info_from_id` over there since it doesn't exist in `account`

File: addons/account/views/account_move_views.xml
Comment (william-andre) [reply addons/account/views/account_move_views.xml:1668]: My issue is that there is a lot of duplicated data

File: addons/account/models/account_journal_dashboard.py
Comment (Megaaaaaa) [reply addons/account/models/account_journal_dashboard.py:463]: Well, it does. For some journals, it seems that the aml query doesn't get the right totals while the move one does. I probably don't know enough about specific use cases to understand why but a test has been written for such a case and without that backup, the sum_waiting and sum_late end up being wrong.

File: addons/account/views/account_move_views.xml
Comment (Megaaaaaa) [reply addons/account/views/account_move_views.xml:1668]: Could you tell me how to use an existing action as a menuitem then ?

Comment (william-andre) [reply addons/account/views/account_move_views.xml:1668]: That you can't do.
On the other hand, you can read the xml-defined action in python and reuse it like this
```python
action = self.env['ir.actions.actions']._for_xml_id('account.action_open_payment_items')
```

File: addons/web/static/src/search/search_bar/search_bar.js
Comment (sdegueldre) [reply addons/web/static/src/search/search_bar/search_bar.js:1]: @Polymorphe57

File: addons/web/static/src/search/search_bar/search_bar.js
Comment (sdegueldre) [inline addons/web/static/src/search/search_bar/search_bar.js:236]: I know this line was just moved but if this works it works by accident, the terms cannot be extracted from this, they should be translated individually
```suggestion
        return ["date", "datetime"].includes(fieldType) ? _t("at") : _t("for");
```

File: addons/web/static/src/search/search_bar/search_bar.js
Comment (william-andre) [inline addons/web/static/src/search/search_bar/search_bar.js:246]: Isn't this cleaner? To potentially avoid playing with indexes
```suggestion
        return {isFieldProperty, fieldType}
```

File: addons/account/models/account_journal_dashboard.py
Comment (william-andre) [inline addons/account/models/account_journal_dashboard.py:465]: So you don't compute anything for the sales journals if you have purchase journals?

File: addons/account/models/account_move_line.py
Comment (william-andre) [inline addons/account/models/account_move_line.py:1149]:
```suggestion
        return [
            '|', 
            '&', ('discount_date', '>=', str(date.today())), ('discount_date', '<=', value), 
            '&', ('discount_date', '<', str(date.today())), ('date_maturity', '<=', value),
        ]
```

We need to search on either one either the other field, so the clauses on today need to be opposite to each other.
Also need to check if it is working as expected when `discount_date` is not set?

Also, add a test case for this (create multiple lines with various dates), and do multiple searches with mocked today (use `freeze_time`)

File: addons/account/models/account_journal_dashboard.py
Comment (william-andre) [reply addons/account/models/account_journal_dashboard.py:468]: > it seems that the aml query doesn't get the right totals while the move one does

What if the query returns an amount that is not right but that is not zero then?

It's the concept of the `or` that makes me think it has to be wrong. It is right or it is not. Maybe you might need to sum both or something like that?

Anyways, needs a test.

File: addons/account/models/account_journal_dashboard.py
Comment (william-andre) [inline addons/account/models/account_journal_dashboard.py:741]: This is still not defined in community

Does any extension from here (the `options` management) actually need to be in community?
It all seems only related to `account_reports`, keep it in `account_reports`

File: addons/account/models/account_journal_dashboard.py
Comment (Megaaaaaa) [reply addons/account/models/account_journal_dashboard.py:468]: If the query retrieved a value, it is right. If it didn't, the other query got it right. Depending on the journal, the aml query will retrieve something or nothing at all. The final result is not some pieces of information coming from one query and some from the other. It is either one or the other, there is no "merge" of data with that `or`. 
I could replace:
```py
for journal in sale_purchase_journals:
            if journal.type == "purchase":
                query_results_to_pay[journal.id] = query_results_payments_to_pay[journal.id] or query_results_bills_to_pay[journal.id]
            else:
                query_results_to_pay[journal.id] = query_results_bills_to_pay[journal.id]
```
by
```py
for journal in sale_purchase_journals:
            if journal.type == "purchase" and query_results_payments_to_pay:
                query_results_to_pay[journal.id] = query_results_payments_to_pay[journal.id]
            else:
                query_results_to_pay[journal.id] = query_results_bills_to_pay[journal.id]
```
if it makes more sense to you.

For the testing part, there is already a test checking the values on the dashboard in odoo/addons/account/tests/test_account_journal_dashboard.py. I didn't modify it because the behavior did not change, We still check the total amount of the bills that are waiting for payment/overdue and those amount don't change based on whether you look at the moves or the amls. The only part of the behavior that is modified is the following : Previously, we were counting the number of moves that were involved in the total and now we count the number of amls that are involved. I can add a test for that part but it seems to me that if the total amount is right, then it means that every line that needed to be taken into account has been so testing the total is sufficient. Now that you know, I let it be your call.

File: addons/account/tests/test_account_payment_items.py
Comment (william-andre) [inline addons/account/tests/test_account_payment_items.py:124]: This is already a nice test :+1: , but what I believe would make it more readable is something like
```python
# bill:               due 2023-03-20
# late_bill:          due 2023-03-01
# discount_bill:      due 2023-03-25 discount 2023-03-20
# late_discount_bill: due 2023-04-01 discount 2023-03-10 
for today, search, expected in [
    ('2023-03-05', '2023-03-01', late_bill),
    ('2023-03-05', '2023-03-30', bill + late_bill + discount_bill + late_discount_bill),
    ...
]:
    with freeze_time(today):
        self.assertEqual(self.env['account.move.line'].search([
            ('payment_date', '=', search),
            ('partner_id', '=', self.private_partner_id.id),
        ]).move_id, expected)
```

Instead of the comments, a helper function that takes those as arguments would be even better
```python
bill = create_bill(due='2023-03-20')
late_bill = create_bill(due='2023-03-01')
discount_bill = create_bill(due='2023-03-25', discount='2023-03-20')
late_discount_bill = create_bill(due='2023-04-01', discount='2023-03-10')
```


This way all the information is at the same place and I think the comments become useless.

_____

In general, always [Keep it DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)

**Comment by william-andre** [comment]
Also, AL asked me to change the commit title from `[IMP] *:` to `[IMP] account:`. It is true that this is mainly a change for `account`. `*` would be used for a more frameworky change

**Comment by Megaaaaaa** [comment]
No change on the enterprise commit's name ? Or same but with account_report ?

**Comment by william-andre** [comment]
> No change on the enterprise commit's name ? Or same but with account_report ?

Same of course

File: addons/account/models/account_journal_dashboard.py
Comment (william-andre) [inline addons/account/models/account_journal_dashboard.py:718]: You still need to filter on the right journal, otherwise you see the lines of all journals.

I think this can be done easily by specifying `action_name` in the context, and using `open_action` instead (see `_select_action_to_open`)

File: addons/account/views/account_journal_dashboard_view.xml
Comment (william-andre) [reply addons/account/views/account_journal_dashboard_view.xml:324]: So you can put it back :)

**Review by william-andre — COMMENTED** [review]
We're getting there! :smile:

**Comment by william-andre** [comment]
You also need to rebase (conflict in `hr_expense`) and fix the CI/style

**Review by william-andre — APPROVED** [review]
@robodoo r+

Thanks!
Sorry if my reviews were a bit hard sometimes, I had a lot to do and not a lot of time to waste :wink:

**Comment by william-andre** [comment]
@robodoo r+

**Pull request closed without merging**
